### ISSUE TITLE
throw error if not all expects are consumed

### ISSUE THREAD

**Issue by piranna** [jprichardson/node-suppose#9]
On https://github.com/jprichardson/node-suppose/blob/12a9044f5effab71521292360e74edcac0f3d88e/lib/suppose-process.js#L89 the callback is called with the exit code, but if there are some expects that are not consumed, nobody knows. Change the callback signature to be Node.js-like being the first argument ```error``` and notify the user that there are some expects not consumed. This can be helpful for testing.

**Comment by jprichardson** [comment]
Do you think it'd be better to instead `emit('error')` somewhere... maybe both?

**Comment by piranna** [comment]
I've been thinking about the ```emit('error')``` think and don't believe it's possible due to two things:

1. we are using the ```.on()``` method to define the expectations, that would conflict with the EventEmitter API (you couldn't know if ```on('error', callback)``` would be for the 'error' event or for the 'error' input string...).
2. you will not be able to know if there are more pending expectations until the process has finished / stream has ended and there will be no more data, so only place to get this "expectations not fullfilled" error is on the callback of the ```.end()``` method.

I know this would break the API, so could be left for a future major version, or maybe do some tricks giving a parameter on the constructor to enable this feature or something...

**Comment by piranna** [comment]
I've found a workaround: expects are available as an attribute, so you can read them on the .end() callback. I think they should be private, but it's an alternative.

**Comment by jprichardson** [comment]
> we are using the .on() method to define the expectations, that would conflict with the EventEmitter API (you couldn't know if on('error', callback) would be for the 'error' event or for the 'error' input string...).

Excellent point. Maybe choosing `on()` was a bad idea. I have no problem breaking the API for the sake of progress. Thoughts on alternative name for `on()`?

**Comment by piranna** [comment]
> Thoughts on alternative name for on()?

.expect()? :-)

I don't see so bad the usage of .on(), it makes sense in that context, but
I would definitely improve it to allow to set several actions for each
expect or none of them (so you would be able to wait for some particular
inputs before doing something).

**Comment by piranna** [comment]
I think this morning, what about not using a Promise like API? :-)

```Javascript
suppose(command, argv)
  .when('hello').then('world')
  .when('2+2').then(function(){return 4})
```

I think this can solve the problem of raising errors by using the Node.js events `.on()` method and also gets a clean and explicit API :-)

**Comment by piranna** [comment]
Fixed in 0.4.0, clossing.
